### PR TITLE
AVI-39: User Authentication | Implement signup functionality

### LINKED ISSUE #39: User Authentication | Implement signup functionality
As a new user, I should be able to register a new account profile.

### TODOs:

- [ ] Consume the API to signup a new user account
- [ ] Make use of Fetch API for HTTP request
- [ ] Store access token using cookies storage (**NOTE**: Not local storage)
- [ ] Upon successful signup, redirect them to the Dashbaord page
- [ ] Implement a guard to protect all the dashboard pages in such a way that it is only accessible by authenticated users (users with an `accessToken`)

### API Details

**API Domain:** https://afrihub-backend.onrender.com
**Signup API:** `POST` https://afrihub-backend.onrender.com/api/auth/signup

**Request Body:**
| Key  | Type |
| ---- | ----- |
| firstName  | string  |
| lastName | string  |
| email  | string  |
| password | string  |

**Response Payload**
```
{ accessToken: '__ACCESS_TOKEN__' }
```

### PR DESCRIPTION
**Pull request by Doyin02**
### What does this PR do

 Implemented user signup functionality using React and Redux on the front end.
 Utilized access tokens for secure user authentication
 Improved error handling for signup, including validation checks and informative error messages.

Closes #39

### How should it be tested

Spin up the app in dev mode with npm start
Head over to the registration 

### Screenshots

Screenshot (197)

### CODE REVIEW

File: src/feature/Registration/index.jsx [hunk 1]
```diff
@@ -1,22 +1,63 @@
...
+  const dispatch = useDispatch();
   const navigate = useNavigate();
 
-  const validateConfirmPassword = (password, confirmPassword) => {
-    return password === confirmPassword;
+  useEffect(() => {
+    if (registrationFailed) {
+      const timeoutId = setTimeout(() => {
+        setRegistrationFailed(false);
+      }, 3000);
```
Comment (harry-olabs) [inline src/feature/Registration/index.jsx:66]: Can you explain why we setting a timeout if we only need to show error alert?

File: src/feature/Registration/index.jsx [hunk 2]
```diff
@@ -35,23 +76,64 @@ const Registration = () => {
...
+        cookies.set("accessToken", accessToken, { path: "/" });
+        console.log(response);
+
+        dispatch(registrationSuccess(userCredentials));
+        navigate("/dashboard");
+      } else {
+        setRegistrationFailed(true);
+      }
+    } catch (error) {
+      console.error("Error:", error);
```
Comment (harry-olabs) [inline src/feature/Registration/index.jsx]: Do not keep console log statement in code. Remove it

File: src/feature/Registration/index.jsx [hunk 3]
```diff
@@ -35,23 +76,64 @@ const Registration = () => {
...
+        }
+      );
+
+      if (response) {
+        const data = await response.json();
+        const { accessToken } = data;
+
+        const cookies = new Cookies();
+        cookies.set("accessToken", accessToken, { path: "/" });
+        console.log(response);
```
Comment (harry-olabs) [inline src/feature/Registration/index.jsx]: Do not keep console log statement in code. Remove it

File: src/index.js [hunk 4]
```diff
@@ -4,12 +4,16 @@ import { BrowserRouter } from "react-router-dom";
 import "./index.css";
 import App from "./App";
 import reportWebVitals from "./reportWebVitals";
+import { Provider } from "react-redux";
+import store from "../src/redux/store";
```
Comment (harry-olabs) [inline src/index.js]: Why did you traverse outside the `src` folder?
Isn't this supposed to be

```suggestion
import store from "./redux/store";
```

File: src/redux/actionTypes.js
Comment (harry-olabs) [inline src/redux/actionTypes.js:1]: The same way you have an `actions` folder, create a `types` folder with every action type files.
For example:
```
redux/types/auth-type.js
```

File: src/redux/actions/authActions.js
Comment (harry-olabs) [inline src/redux/actions/authActions.js:1]: For the file name, use the naming convention with hyphen
For example:
```
auth-actions.js
```

File: src/redux/rootReducer.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { combineReducers } from "redux";
+import authReducer from "../../src/redux/reducers/authReducer";
+
+const rootReducer = combineReducers({
+  auth: authReducer,
+});
+
+export default rootReducer;
```
Comment (harry-olabs) [inline src/redux/rootReducer.js]: Create an `index.js` file in the `reducers` folder and move this codes in there

**Review by harry-olabs — CHANGES_REQUESTED** [review]
@Doyin02 it is looking good, and some adjustments. Attend to the comments

File: src/redux/actionTypes.js
Comment (Doyin02) [reply src/redux/actionTypes.js:1]: I don't understand sir

File: src/redux/store.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { configureStore, combineReducers } from "@reduxjs/toolkit";
+import authReducer from "./reducers/auth-reducer";
+
+const rootReducer = combineReducers({
+  auth: authReducer,
+});
```
Comment (harry-olabs) [inline src/redux/store.js:6]: This is already already defined in the `reducers/index` file — remove it 🔥

File: src/utils/validation.js [hunk 7]
```diff
@@ -6,6 +6,10 @@ export const validateEmail = (email) => {
   export const validatePassword = (password) => {
     return password.length >= 6;
   };
+
+    export const validateConfirmPassword = (password, confirmPassword) => {
+    return password === confirmPassword;
+  };
```
Comment (harry-olabs) [inline src/utils/validation.js:12]: Format this properly

File: src/App.js [hunk 8]
```diff
@@ -1,5 +1,7 @@
 import React from "react";
 import { Routes } from "./routes";
+import 'bootstrap/dist/css/bootstrap.min.css';
+import 'bootstrap/dist/js/bootstrap.min.js';
```
Comment (harry-olabs) [inline src/App.js:4]: Why don't you add this to the `src/index.css` file instead?

File: src/feature/Registration/index.jsx [hunk 9]
```diff
@@ -35,23 +76,62 @@ const Registration = () => {
...
+          method: "POST",
+          headers: {
+            "Content-Type": "application/json",
+          },
+          body: JSON.stringify(userCredentials),
+        }
+      );
+
+      if (response) {
+        const data = await response.json();
```
Comment (harry-olabs) [inline src/feature/Registration/index.jsx:119]: Start adopting custom hooks for http requests and services.
Follow this tutorial to refactor this into a custom hook `useFetch`.

https://blog.openreplay.com/building-a-custom-fetch-hook-in-react/#:~:text=Building%20the%20custom%20Fetch%20Hook&text=You%20can%20name%20the%20hook,are%20fetching%20the%20data%20from.

File: src/feature/Registration/index.jsx [hunk 10]
```diff
@@ -1,22 +1,63 @@
...
+  validateConfirmPassword,
+} from "../../utils/validation";
+import Cookies from "universal-cookie";
 
 const Registration = () => {
+  const [firstName, setFirstName] = useState("");
+  const [lastName, setLastName] = useState("");
   const [email, setEmail] = useState("");
   const [password, setPassword] = useState("");
   const [confirmPassword, setConfirmPassword] = useState("");
```
Comment (harry-olabs) [inline src/feature/Registration/index.jsx:21]: Refactor this into a single state:

```suggestion
  const signupForm = useState({
    firstName: '',
    lastName: '',
    email: '',
    password: '',
    confirmPassword: '',
  });
```

For each input element, give it the matching name key. For example:
```
<input name="firstName" onChange={setFormValue} />
```

Then to update form value, simple do:

```js
const setFormValue = (el) => {
    setSignupForm(form => ({
      ...form,
      [el.target.name]: el.target.value
    }))
  }
```

After these, clean up the code to remove all the repetative handleChange function.

**Review by harry-olabs — CHANGES_REQUESTED** [review]

File: src/feature/Registration/index.jsx [hunk 11]
```diff
@@ -1,57 +1,164 @@
...
+  validateConfirmPassword,
+} from "../../utils/validation";
+import Cookies from "universal-cookie";
+import useFetch from "../../hooks/use-fetch";
 
 const Registration = () => {
-  const [email, setEmail] = useState("");
-  const [password, setPassword] = useState("");
-  const [confirmPassword, setConfirmPassword] = useState("");
+  const [state, setState] = useState({
```
Comment (harry-olabs) [inline src/feature/Registration/index.jsx]: Use name like `formData` and `setFormData` instead.

File: src/hooks/use-fetch-bak.js
Comment (harry-olabs) [inline src/hooks/use-fetch-bak.js:1]: This is a duplicate file and is not being used? Delete it 🔥

**Review by harry-olabs — CHANGES_REQUESTED** [review]
It is looking good, just minor changes

**Review by harry-olabs — APPROVED** [review]

**Pull request merged**